### ISSUE TITLE
6. Regression_sklearn_regression

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#28]
"""
sklearn 관련 Linear Regression
"""

from sklearn.linear_model import LinearRegression # 회귀모델의 객체를 만들수 있다.# model object 생성
from sklearn.model_selection import train_test_split # train/test split
from sklearn.metrics import mean_squared_error, r2_score # model 평가
from sklearn.datasets import load_diabetes # dataset
import numpy as np # 숫자처리
import pandas as pd # data.frame, 상관계수
import matplotlib.pyplot as plt # 회귀선 시각화

### 1. dataset load
X, y = load_diabetes(return_X_y = True) # 엑스와이를 동시에 리턴받기
X.shape # (442, 10)
y.shape # (442,)

y.mean() # 152.13348416289594

### 2. x,y 변수
-x(bmi : 비만도지수) -> y

x_bmi = X[:,2]
x_bmi.shape # (442,)

### 3. model 생성 : object -> traing -> model
obj = LinearRegression() # 생성자 object
model = obj.fit(x_bmi, y) #(X, y) -> model
-오류
-x_bmi가 변수가 1개여도 2차원의 형태로 바꾸고 모델에 넣어주어야 한다.

- 1d -> 2d
x_bmi = x_bmi.reshape(442, 1)
model = obj.fit(x_bmi, y)
model

- 예측치
y_pred = model.predict(x_bmi) # predict(X)
y_pred.shape # (442,)
y.shape # (442,)

### 4. model 평가 : MSE(y가 정규화된 경우), r2_score(비정규화된 경우) 
MSE = mean_squared_error(y,y_pred) # (y 정답, y 예측치)
MSE # 3890.4565854612724
score = r2_score(y, y_pred) # (y 정답, y 예측치)
score # 0.3439237602253803 - 1에 가까울수록 정답과 예측치차이가 적다.

### 5. dataset split(70:30)
-전까지는 모든 데이터를 학습시키고, 모든 데이터를 평가했는데,
-이번에는 훈련과 테스트셋을 나누어서 모델을 만들고 평가해보자.
x_train, x_test, y_train, y_test = train_test_split(x_bmi, y, test_size = 0.3)
x_train.shape # (309, 1)
x_test.shape # (133, 1)
y_train.shape # (309,)
y_test.shape # (133, 1)

- model 생성
obj = LinearRegression() # 생성자 object
model = obj.fit(x_train,y_train) # training dataset

y_pred = model.predict(x_test) # test dataset이용하여 예측치 구하기

- model 평가 : MSE(y가 정규화된 경우), r2_score(비정규화된 경우) 
MSE = mean_squared_error(y_test,y_pred) # (y 정답, y 예측치)
MSE # 3730.2071378632186
score = r2_score(y_test, y_pred) # (y 정답, y 예측치)
score # 0.3272661540910551 - 0에 가까울수록 정답과 예측치차이가 적다.

y_test[:10]
y_pred[:10]

import pandas as pd
df = pd.DataFrame({'y_true': y_test, 'y_pred':y_pred})
cor = df['y_true'].corr(df['y_pred'])
cor # 0.5763770475322645 숫자가1에 가까울수록 예측을 잘 한 것

import matplotlib.pyplot as plt # 회귀선 시각화
plt.plot(x_test, y_test, 'ro') # 산점도
plt.plot(x_test,y_pred, 'b-') # 회귀선
plt.show()

### 1. dataset load 
iris = pd.read_csv("C:\\ITWILL\\4_Python-II\\data/iris.csv")
iris.info()
'''
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  150 non-null    float64 ->y
 1   Sepal.Width   150 non-null    float64 ->x1
 2   Petal.Length  150 non-null    float64 ->x2
 3   Petal.Width   150 non-null    float64 ->x3
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
'''

### 2. x,y 변수 선택
cols = list(iris.columns)  
cols # ['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width', 'Species']

y_col = cols[0] # 'Sepal.Length'
x_cols = cols[1:-1] # 'Sepal.Width', 'Petal.Length', 'Petal.Width'
x_cols
y_col

### 3. dataset split
iris_train, iris_test = train_test_split(iris, test_size =0.3, random_state=123) # default size =0.25
'''
test_size : 검정데이터셋 비율(default = 0.25)
random_state : sampling seed 값(항상 동일한 데이터가 샘플링되도록)
'''

iris_train.shape  # (105, 5)
iris_test.shape # (45, 5)

iris_train.head()
iris_test.head()

### 4. model 생성 : train data 이용하여
lr = LinearRegression()
model = lr.fit(X= iris_train[x_cols],y=iris_train[y_col])
model

### 5. model 평가 : test data 이용하여
y_pred = model.predict(X=iris_test[x_cols]) # 예측치
y_true = iris_test[y_col] # 관측치(정답) = label

y_true.min() # 4.3
y_true.max() # 7.9

-어느 정도 정규화가 되어서 mse로 평가하자.
-평균제곱오차 : mean((y_true-y_pred)**2)
mse = mean_squared_error(y_true, y_pred) # 0.11633863200224723
-0에 가까울수록
score = r2_score(y_true, y_pred) # 0.8546807657451759
-1에 가까울수록/ 85% 예측력

- y_true vs y_pred 시각화
-타입 맞추어 주기/ 안맞추니 이상하게 그래프 나옴
type(y_true) # pandas.core.series.Series
y_true = np.array(y_true)
type(y_pred) # numpy.ndarray

import matplotlib.pyplot as plt
fig = plt.figure(figsize = (10,5))
chart = fig.subplots()
chart.plot(y_true, color = 'b', label = 'real values')
chart.plot(y_pred, color = 'r', label = 'fitted values') # 예측치
plt.legend(loc='best')
plt.show()
